# Воронки
Попытайтесь понять и обосновать, как меняется воронка по месяцам, если учесть еще и пол посетителей. 
Загрузите данные кликстрима из файла click_stream3.csv.
В данных появился еще один новый столбец - Gender.
Определите, влияет ли пол посетителя на какую-либо воронку какого-либо месяца на каком-то устройстве?

In [68]:
# Разделение данных из файла click_stream2.csv с учетом устройств
import csv

funnel_by_month_mobile = {}
funnel_by_month_desktop = {}
funnel_template = {'1_home_page':0, '2_search_page':0, '3_payment_page':0, '4_payment_confirmation_page':0}

with open('click_stream2.csv', mode='r') as csv_file:
    csv_reader = csv.DictReader(csv_file, fieldnames=['ID','page','date','device'])

    for row in csv_reader: 
        
        page = list(row.items())[1][1]
        event_date = list(row.items())[2][1][:-3]
        device = list(row.items())[3][1]
            
        if device == 'Mobile':
            if event_date not in funnel_by_month_mobile:
                funnel_by_month_mobile[event_date] = funnel_template.copy()
            
            if page == '1_home_page':
                funnel_by_month_mobile[event_date]['1_home_page'] += 1
            elif page == '2_search_page':
                funnel_by_month_mobile[event_date]['2_search_page'] += 1
            elif page == '3_payment_page':
                funnel_by_month_mobile[event_date]['3_payment_page'] += 1
            else:
                funnel_by_month_mobile[event_date]['4_payment_confirmation_page'] += 1   
        elif device == 'Desktop':       
            if event_date not in funnel_by_month_desktop:
                    funnel_by_month_desktop[event_date] = funnel_template.copy()
            if page == '1_home_page':
                funnel_by_month_desktop[event_date]['1_home_page'] += 1
            elif page == '2_search_page':
                funnel_by_month_desktop[event_date]['2_search_page'] += 1
            elif page == '3_payment_page':
                funnel_by_month_desktop[event_date]['3_payment_page'] += 1
            else:
                funnel_by_month_desktop[event_date]['4_payment_confirmation_page'] += 1

print("Данные кликстрима по месяцам, мобильные:")
keys_funnel_by_month_mobile = list(funnel_by_month_mobile.keys())
keys_funnel_by_month_mobile.sort()
for i in keys_funnel_by_month_mobile:
    print(i, ':', funnel_by_month_mobile[i])            
print("Данные кликстрима по месяцам, десктопы:")
keys_funnel_by_month_desktop = list(funnel_by_month_desktop.keys())
keys_funnel_by_month_desktop.sort()
for i in keys_funnel_by_month_desktop:
    print(i, ':', funnel_by_month_desktop[i])

Данные кликстрима по месяцам, мобильные:
2015-01 : {'1_home_page': 7550, '2_search_page': 6025, '3_payment_page': 1169, '4_payment_confirmation_page': 129}
2015-02 : {'1_home_page': 7550, '2_search_page': 6055, '3_payment_page': 1221, '4_payment_confirmation_page': 119}
2015-03 : {'1_home_page': 7550, '2_search_page': 1509, '3_payment_page': 335, '4_payment_confirmation_page': 30}
2015-04 : {'1_home_page': 7550, '2_search_page': 1511, '3_payment_page': 295, '4_payment_confirmation_page': 24}
Данные кликстрима по месяцам, десктопы:
2015-01 : {'1_home_page': 15050, '2_search_page': 7529, '3_payment_page': 1221, '4_payment_confirmation_page': 60}
2015-02 : {'1_home_page': 15050, '2_search_page': 7632, '3_payment_page': 1191, '4_payment_confirmation_page': 54}
2015-03 : {'1_home_page': 15050, '2_search_page': 7370, '3_payment_page': 296, '4_payment_confirmation_page': 14}
2015-04 : {'1_home_page': 15050, '2_search_page': 7569, '3_payment_page': 302, '4_payment_confirmation_page': 22}


In [69]:
# Разделение данных из файла click_stream3.csv с учетом устройств и пола
import csv

funnel_by_month_mobile_male = {}
funnel_by_month_mobile_female = {}
funnel_by_month_desktop_male = {}
funnel_by_month_desktop_female = {}
funnel_template = {'1_home_page':0, '2_search_page':0, '3_payment_page':0, '4_payment_confirmation_page':0}

with open('click_stream3.csv', mode='r') as csv_file:
    csv_reader = csv.DictReader(csv_file, fieldnames=['ID','page','date','device','gender'])

    for row in csv_reader: 
        
        page = list(row.items())[1][1]
        event_date = list(row.items())[2][1][:-3]
        device = list(row.items())[3][1]
        gender = list(row.items())[4][1]
            
        if device == 'Mobile':
            if gender == 'Male':
                if event_date not in funnel_by_month_mobile_male:
                    funnel_by_month_mobile_male[event_date] = funnel_template.copy()
                if page == '1_home_page':
                    funnel_by_month_mobile_male[event_date]['1_home_page'] += 1
                elif page == '2_search_page':
                    funnel_by_month_mobile_male[event_date]['2_search_page'] += 1
                elif page == '3_payment_page':
                    funnel_by_month_mobile_male[event_date]['3_payment_page'] += 1
                else:
                    funnel_by_month_mobile_male[event_date]['4_payment_confirmation_page'] += 1
            elif gender == 'Female':
                if event_date not in funnel_by_month_mobile_female:
                    funnel_by_month_mobile_female[event_date] = funnel_template.copy()
                if page == '1_home_page':
                    funnel_by_month_mobile_female[event_date]['1_home_page'] += 1
                elif page == '2_search_page':
                    funnel_by_month_mobile_female[event_date]['2_search_page'] += 1
                elif page == '3_payment_page':
                    funnel_by_month_mobile_female[event_date]['3_payment_page'] += 1
                else:
                    funnel_by_month_mobile_female[event_date]['4_payment_confirmation_page'] += 1
        elif device == 'Desktop':
            if gender == 'Male':
                if event_date not in funnel_by_month_desktop_male:
                    funnel_by_month_desktop_male[event_date] = funnel_template.copy()
                if page == '1_home_page':
                    funnel_by_month_desktop_male[event_date]['1_home_page'] += 1
                elif page == '2_search_page':
                    funnel_by_month_desktop_male[event_date]['2_search_page'] += 1
                elif page == '3_payment_page':
                    funnel_by_month_desktop_male[event_date]['3_payment_page'] += 1
                else:
                    funnel_by_month_desktop_male[event_date]['4_payment_confirmation_page'] += 1
            elif gender == 'Female':
                if event_date not in funnel_by_month_desktop_female:
                    funnel_by_month_desktop_female[event_date] = funnel_template.copy()
                if page == '1_home_page':
                    funnel_by_month_desktop_female[event_date]['1_home_page'] += 1
                elif page == '2_search_page':
                    funnel_by_month_desktop_female[event_date]['2_search_page'] += 1
                elif page == '3_payment_page':
                    funnel_by_month_desktop_female[event_date]['3_payment_page'] += 1
                else:
                    funnel_by_month_desktop_female[event_date]['4_payment_confirmation_page'] += 1                
                
print("Данные кликстрима по месяцам, мобильные + мужчины:")
keys_funnel_by_month_mobile_male = list(funnel_by_month_mobile_male.keys())
keys_funnel_by_month_mobile_male.sort()
for i in keys_funnel_by_month_mobile_male:
    print(i, ':', funnel_by_month_mobile_male[i])
print("Данные кликстрима по месяцам, мобильные + женщины:")
keys_funnel_by_month_mobile_female = list(funnel_by_month_mobile_female.keys())
keys_funnel_by_month_mobile_female.sort()
for i in keys_funnel_by_month_mobile_female:
    print(i, ':', funnel_by_month_mobile_female[i])  
print("Данные кликстрима по месяцам, десктопы + мужчины:")
keys_funnel_by_month_desktop_male = list(funnel_by_month_desktop_male.keys())
keys_funnel_by_month_desktop_male.sort()
for i in keys_funnel_by_month_desktop_male:
    print(i, ':', funnel_by_month_desktop_male[i])         
print("Данные кликстрима по месяцам, десктопы + женщины:")
keys_funnel_by_month_desktop_female = list(funnel_by_month_desktop_female.keys())
keys_funnel_by_month_desktop_female.sort()
for i in keys_funnel_by_month_desktop_female:
    print(i, ':', funnel_by_month_desktop_female[i])

Данные кликстрима по месяцам, мобильные + мужчины:
2015-01 : {'1_home_page': 3682, '2_search_page': 2963, '3_payment_page': 551, '4_payment_confirmation_page': 53}
2015-02 : {'1_home_page': 3799, '2_search_page': 3035, '3_payment_page': 591, '4_payment_confirmation_page': 59}
2015-03 : {'1_home_page': 3812, '2_search_page': 762, '3_payment_page': 160, '4_payment_confirmation_page': 10}
2015-04 : {'1_home_page': 3829, '2_search_page': 755, '3_payment_page': 148, '4_payment_confirmation_page': 13}
Данные кликстрима по месяцам, мобильные + женщины:
2015-01 : {'1_home_page': 3868, '2_search_page': 3062, '3_payment_page': 618, '4_payment_confirmation_page': 76}
2015-02 : {'1_home_page': 3751, '2_search_page': 3020, '3_payment_page': 630, '4_payment_confirmation_page': 60}
2015-03 : {'1_home_page': 3738, '2_search_page': 747, '3_payment_page': 175, '4_payment_confirmation_page': 20}
2015-04 : {'1_home_page': 3721, '2_search_page': 756, '3_payment_page': 147, '4_payment_confirmation_page': 11

**Комментарии:** 
<br>Согласно полученным данным и визуализации данных в приложенном файле пол не влияет на воронки, так как процент конверсий каждого шага в каждой воронке в зависимости от пола сильно не отличается. На средних шагах (2_search_page и 3_payment_page) разница составляет 1-2%, а на последнем шаге разница доходит до 5% (6% - мужчины, 11% - женщины) только в марте в случае использования мобильных устройств, что можно определить как погрешность из-за огромного скачка конверсий в целом (например, закрытие отложенных корзин, увеличение спроса женщинами в праздничный период и т.д.)<br>
<br>P.S.<br>
<br>Несмотря на то, что входящий трафик с мобильных устройств почти в два раза меньше, покупок больше именно с них. Особенно это заметно в первые два месяца до поломок средних страниц сайта, когда продажи с мобильных устройств превышают болеее, чем в два раза продажи с десктопных. То есть мобильная версия сайта как минимум в четыре раза эффективнее, что можно учитывать при маркетинговых коммуникациях (например, показывать контекстную рекламу только для мобильных телефонов). С учетом низкой конверсии на декстопах с домашней страницы на вторую 2_search_page в сравнении с таким же показателем на мобильных устройствах, кроме технических исправлений, необходимо привлекать пользователей сразу на целевые страницы среднего уровня, что требует более тонкой настройки рекламных кампаний в Интернете.<br>